### ISSUE TITLE
useInfiniteQuery does not continue to fetch after first load

### ISSUE THREAD

**Issue by johnslay** [tannerlinsley/react-query#849]
**Describe the bug**
Just upgraded from 2.4.15 to 2.5.13 and noticed a breaking change in my application.

When using `useInfiniteQuery` and immediately calling `fetchMore` to exhaust the list of resources, the request is skipped due to `this.promise` stuck in pending state. See example below:

```ts
const useListClusters = () => {
  const query = useInfiniteQuery<ClustersQuery, string, string>(
    'listClusters',
    async (_, prevNextToken?: string) => {
      const { clusters, nextToken } = await listClusters({ nextToken: prevNextToken });
      return { clusters, nextToken };
    },
    { getFetchMore: (last) => last.nextToken }
  );

  return query;
};
```

Usage:
```ts
const CardView = () => {
  const { data, canFetchMore } = useListClusters();

  React.useEffect(() => {
    if (canFetchMore) {
      fetchMore();
    }
  });

  const clusters = React.useMemo<string[]>(() => {
    return data?.reduce((acc, curr) => acc.concat(curr.clusters), []) ?? [];
  }, [data]);

  return (
    <View data={clusters}>...</View>
  );
};
```

After tracing the call with the debugger, I have found that the issue is because the promise is still stuck in the `pending` state even though it has completed and the ui has updated.

[/src/core/query.ts#L397-L399](https://github.com/tannerlinsley/react-query/blob/58acca1c6b07847c5b120b7c888aa088bdd1232b/src/core/query.ts#L397-L399):

**To Reproduce**
1. Use a paginated api
2. Create example component from above.
3. Create enough resource for the call to return a `nextToken`
4. See only 1 call get made even when `nextToken` is returned.

https://codesandbox.io/s/react-query-test-ebsco

**Expected behavior**
If `nextToken` is returned, I should see another call to the api.

**Desktop (please complete the following information):**
 - OS: MacOS Mojave
 - Browser: chrome
 - Version: 84

**Additional context**
The functionality was broken shortly after 2.4.15 when the migration to TS occurred.

**Comment by johnslay** [comment]
Another interesting bug somewhat related to this:

So I have about 16 pages worth of data. After the page mounts, the current page count in the query is 1 (due to the bug above). After the page is settled, I click a button that calls `refetch`. In the network tab, I see the correct amount of queries (16 total). However, the query object only shows 9 and the total list of items is reflected inaccurately on the page. 

Seems like there is some sort of race condition going on here causing some queries to be skipped or overlooked?

**Comment by johnslay** [comment]
So I've hacked up an environment to repro this issue here: https://codesandbox.io/s/react-query-test-ebsco

However, as you'll see, it works just fine...

So I went back to my environment to test this out and come to find out that the query stops after the first request depending on what I render in the return statement..

If I just return something like:
```
return (
      <ul>
        {clusters.map((i) => (
          <li key="i">{i}</li>
        ))}
      </ul>
);
```

Everything works just fine. I see 16 network calls in the network tab.

In my case I have a component `UnifiedLayout`

and where it breaks is the moment I render something like:
```
return (
  <UnifiedLayout>
      <ul>
        {clusters.map((i) => (
          <li key="i">{i}</li>
        ))}
      </ul>
  <UnifiedLayout>
);
```

I see only 1 network call in this case...

Any idea whats going on here?
